### ISSUE TITLE
Import map not working in different browsers

### ISSUE THREAD

**Issue by badde-media** [guybedford/es-module-shims#270]
Firefox and some other browsers (details below) have problems resolving the importmap. Error:
```
Uncaught TypeError: Error resolving module specifier “three”. Relative module specifiers must start with “./”, “../” or “/”.
```
The line it refers to is
```
import { ... } from /*'three'*/'blob:https://[URL]/c945bbba-09f5-454f-b2a4-508d3d70f89a';
```
which is the Firefox translation of
```
import { ... } from 'three';
```
referring to the import map in the index.html file:
```
	<!-- Import maps polyfill -->
	<!-- Remove this when import maps will be widely supported -->
	<script async src="https://unpkg.com/es-module-shims@1.3.6/dist/es-module-shims.js"></script>

	<script type="importmap">
		{
			"imports": {
				"three": "./vendor/three/build/three.module.js"
			}
		}
	</script>
```
This code works perfectly fine in Chrome Canary.

I have tested the [online version](https://badde.media/threejs/fungi/index.html) in Chrome, Chrome Canary, Chrome Mobile Android, Opera, Opera Mobile Android, Edge, and it worked. I’ve tested it in Firefox, Firefox Mobile Android, Chrome Mobile iOS, Safari Mobile, and it didn’t work - although other projects of mine did, like [this one](https://badde.media/threejs/shapes/index.html)

To Reproduce
Try [this link](https://badde.media/threejs/fungi/index.html) in Firefox, Firefox Mobile Android, Chrome Mobile iOS or Safari Mobile.

Go to [this link](https://badde.media/threejs/fungi/index.html)
Page should load
but doesn't. Or maybe partially, with errors like the black background staying black instead of becoming colored .

Code

```
	<!-- Import maps polyfill -->
	<!-- Remove this when import maps will be widely supported -->
	<!-- <script async src="https://unpkg.com/es-module-shims@1.3.6/dist/es-module-shims.js"></script> -->

	<script type="importmap">
		{
			"imports": {
				"three": "./vendor/three/build/three.module.js"
			}
		}
	</script>
	
import { Vector3, Object3D, Mesh, TextureLoader } from 'three';
```
Live example

[jsfiddle-latest-release](https://jsfiddle.net/hyok6tvj/)
[jsfiddle-dev](https://jsfiddle.net/c5m1kazu/)
THESE ^ two examples actually show the same error.

Expected behavior
When the site has loaded, hundred small colored spheres should spread out from the center inside of a black sphere where the camera also is located in. When one of these small spheres hit the big one, the small ones should disappear, the bigsphere should take the color of the hitting sphere (console.logged), and an inner sphere with wireframe shader should appear with the inverted color. In the middle, a patch of grass should occur where a mushroom grows out (mushroom /growing are WIP) Image shows expected behaviour in Opera Mobile browser.
![Screenshot_20220310-103929_Opera](https://user-images.githubusercontent.com/19553316/157935697-ddaa5c87-ec8c-4537-9ae2-39ca82d1afa5.jpg)
.

**Comment by guybedford** [comment]
@badde-media thanks for the clear report. The error you mention is expected because that is the "polyfill error" as described in https://github.com/guybedford/es-module-shims#usage, so it's another issue that the polyfill isn't working. But you're running an old version so this is likely fixed - make sure to use the latest 1.4.7 version - this should definitely be working. Please try upgrading to that and let me know if the issue is still there?

**Comment by badde-media** [comment]
@guybedford Thank you for your reply. Unfortunately, the problem doesn't seem to be solved by using the newer version.
I've changed the line
```
	<script async src="https://unpkg.com/es-module-shims@1.3.6/dist/es-module-shims.js"></script>
```
into 
```
	<script async src="https://unpkg.com/es-module-shims@1.4.7 /dist/es-module-shims.js"></script>
```
and now get the following error messages:
New:
```
asm.js type error: Asm.js optimizer disabled because no suitable wasm compiler is available es-module-shims.js
```
and still:
```
Uncaught TypeError: Error resolving module specifier “three”. Relative module specifiers must start with “./”, “../” or “/”.
```
and another "Uncaught TypeError: undefined has no properties".
For now, I have only tested this on Firefox Desktop v98.0.

**Comment by guybedford** [comment]
@badde-media here is the issue - you have a cycle between the `World.js` file and the `earth.js` file, where `World.js` defines a locally exported live binding `scene`, then calls a function in `earth.js` which expects that live binding to be present.

Due to the way the polyfill works, when dealing with cycles we use a shell module technique that won't propagate live bindings in this exception case - instead you need to ensure that bindings in a cycle are only ever accessed after completion of the module's execution in the cycle.

This is described in the readme here - https://github.com/guybedford/es-module-shims#import-rewriting.

**Comment by badde-media** [comment]
I don't completely understand what this means 
"When executing a circular reference A -> B -> A, a shell module technique is used to "shim" the circular reference into an acyclic graph. As a result, live bindings for the circular parent A are not supported, and instead the bindings are captured immediately after the execution of A."
Can you please show me an example of such an A-B-A setup that works?

**Comment by guybedford** [comment]
Here's a simple example of the situation:

a.js
```js
import { fn } from './b.js';
export let val = 5;
// should log 5
console.log(fn());
```

b.js
```js
import { val } from './a.js';
export function fn () {
  return val;
}
```

Importing `a.js` will execute `b.js` first and then `a.js`. The binding `val` being imported from `a.js` is live so when calling `fn` in `b.js` its value is correct. The problem es-module-shims has with the cycle is that we use blob URLs to polyfill import maps, which are unique for each module. It's impossible to have a cycle this way since you need a unique URL to contain another unique URL known before you got the second unique URL.

So instead ES Module Shims creates a shell model for the cycle:

a.js
```js
import { fn } from './b.js';
import { __updater } from './a-shell.js';
import * as realA from './a.js';
export let val = 5;
console.log(fn());

// this updates the shell only after the execution of a
__updater(realA);
```

b.js
```js
import { val } from './a-shell.js';
export function fn () {
  return val;
}
```

a-shell.js
```js
export let val = undefined;
export function __updater(realA) {
  val = realA.val;
}
```

There could be tricks to eg inject the update function after each statement line of a.js to provide a better live bindings capture during the execution of a. But the recommended approach for es-module-shims is just not to expect live bindings propagation in cycles. Live bindings will work correctly everywhere else though.

**Comment by badde-media** [comment]
Sorry, but I don't get what the benefit of this is, when it breaks things that would usually work. And I don't understand why different browsers behave differently with the same script. JavaScript should work browser-independent, no? Now, I have to set up extra files, functions, imports and exports, just in order to get it running as it worked before, in all browsers. Do I have to make different versions for each browser? You're probably far more experienced than me concerning JS, please explain me why this is necessary.

**Comment by badde-media** [comment]
Yay, I'm the first one, lol. Thanks for your answers, I'll try to implement them. And I'll think of a way how to avoid the cycle.
